### PR TITLE
WIP: Add artwork (thumbnail) URL to AudioTrackInfo

### PR DESCRIPTION
**Pull request by GoldRenard**
This PR will add new `AudioTrackInfo#artworkUri` field containing artwork or thumbnail url of requested resource. It is supported for:
* Bandcamp
* Beam
* Nico (not tested but should be ok)
* SoundCloud
* Twitch
* Vimeo
* Youtube

### CODE REVIEW

**Comment by GoldRenard** [comment]
@sedmelluq please review this.

**Comment by Yucked** [comment]
BUMP.

File: main/src/main/java/com/sedmelluq/discord/lavaplayer/source/soundcloud/SoundCloudAudioSourceManager.java [hunk 1]
```diff
@@ -298,13 +298,15 @@ private AudioTrack loadFromTrackPage(String trackWebUrl, String secretToken) {
...
+    JsonBrowser artworkUrl = trackInfoJson.get("artwork_url");
     AudioTrackInfo trackInfo = new AudioTrackInfo(
         trackInfoJson.get("title").text(),
         trackInfoJson.get("user").get("username").text(),
         trackInfoJson.get("duration").as(Integer.class),
         secretToken != null ? trackId + "|" + secretToken : trackId,
         false,
-        trackInfoJson.get("permalink_url").text()
+        trackInfoJson.get("permalink_url").text(),
+        artworkUrl != null ? artworkUrl.text() : null
```
[the remark this answers is not in the document]
Comment (Shadorc) [reply main/src/main/java/com/sedmelluq/discord/lavaplayer/source/soundcloud/SoundCloudAudioSourceManager.java]: You're doing artworkUrl.text() = artworkUrl.text() with this

[the remark this answers is not in the document]
Comment (Zenrac) [reply main/src/main/java/com/sedmelluq/discord/lavaplayer/source/soundcloud/SoundCloudAudioSourceManager.java]: True thanks

[the remark this answers is not in the document]
Comment (GoldRenard) [reply main/src/main/java/com/sedmelluq/discord/lavaplayer/source/soundcloud/SoundCloudAudioSourceManager.java]: You're right. Removed this unnecessary null check.

**Comment by GoldRenard** [comment]
It's been a year... @sedmelluq please give a comment at least why this PR is still open.

**Comment by MrJacz** [comment]
@sedmelluq Any update on this?

**Comment by GoldRenard** [comment]
@sedmelluq bump

**Comment by JamesParkDev** [comment]
bump

**Comment by GoldRenard** [comment]
and many many bumps again

_\*bumps, bumps, bumps\*_

**Comment by sedmelluq** [comment]
There will be no more hardcoded fields added to `AudioTrackInfo` as it is not a suitable place for holding metadata (aka data that is not required for actual playback). The main cause is that I don't want to add a breaking serialization change for this and for any other metadata change, only changes in the fields that are actually required for playback may change serialization format.

The intended final approach is to turn it into an interface/abstract class with a map for metadata and methods for accessing various metadata which can have per track type implementations (either fetch from map or calculate from other fields).

As an intermediate step I am willing to accept adding a `Map<String, String> metadata` to the current class, storing artwork url in it, and adding a method which provides it from there. However, serialization should not currently be changed (decode would get an empty map), thus artwork would not survive serialize->deserialize.

**Comment by GoldRenard** [comment]
Serialization is an important to keep all data. For example, my implementation of playlists is using custom `AudioSourceManager` which loads all tracks from storage and deserializes them to `AuditTrack` inside custom `BasicAudioPlaylist` implementation. I don't want artwork to be lost in that case.

Having metadata map is great idea (and leaving it modifiable for developers may be useful too), but I really want this metadata to be serializable too in that case. And as far as I can see, bumping track info version will not break anything since there is implicit check for versions. At least if we will bump it once just for metadata, this metadata can be used for anything in future.

**Comment by GoldRenard** [comment]
Well if we will leave this metadata modifiable we can handle such cases.

**Comment by officialpiyush** [comment]
we need this

**Comment by GoldRenard** [comment]
I will recreate this PR soon with fresh branch.

**Pull request closed without merging**